osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs

```diff
@@ -37,23 +35,51 @@ public void ApplyToBeatmap(IBeatmap beatmap)
 
             var positionInfos = OsuHitObjectGenerationUtils.GeneratePositionInfos(osuBeatmap.HitObjects);
 
-            float rateOfChangeMultiplier = 0;
+            float sequenceOffset = 0;
+            bool flowDirection = false;
 
-            foreach (var positionInfo in positionInfos)
+            for (int i = 0; i < positionInfos.Count; i++)
             {
-                // rateOfChangeMultiplier only changes every 5 iterations in a combo
-                // to prevent shaky-line-shaped streams
-                if (positionInfo.HitObject.IndexInCurrentCombo % 5 == 0)
-                    rateOfChangeMultiplier = (float)rng.NextDouble() * 2 - 1;
+                bool invertFlow = false;
+
+                if (i == 0 ||
+                    (positionInfos[i - 1].HitObject.NewCombo && (i <= 1 || !positionInfos[i - 2].HitObject.NewCombo) && (i <= 2 || !positionInfos[i - 3].HitObject.NewCombo)) ||
+                    OsuHitObjectGenerationUtils.IsHitObjectOnBeat(osuBeatmap, positionInfos[i - 1].HitObject, true) ||
+                    (OsuHitObjectGenerationUtils.IsHitObjectOnBeat(osuBeatmap, positionInfos[i - 1].HitObject) && rng.NextDouble() < 0.25))
```

Comment:
There's a lot of randomness peppered in here and below for no real discernible reason. Why was this done? How were the thresholds chosen?

Reply:
I added those because IMO changing the flow on every downbeat would be too much since that's not what most beatmaps look like, so it only changes the flow and angle offsets with a certain probability. I don't really have a reason why I chose these _exact_ values, but after testing and adjusting the thresholds I found that these values generated decent results.